docs: add fabric nested pipeline example (#110)
microsoft/data-factory-testing-framework: examples/fabric/nested_ingestion_pipeline/README.md

# Nested pipelines

The goal of the pipeline is to iterate through a configuration json file and for each iteration item verify if the combination of the year/month data to be ingested is new or have been updated in the source system. If that is the case then a nested pipeline is invoked and will be responsible to execute the ingestion in 2 steps:

- from the web to an adlsgen2 container
- from the adlsgen2 container to the Fabric lakehouse

If the data is already in the lakehouse and has not been updated, the ingestion pipeline is not triggered.

![Nested Pipelines](pl_main_pipeline.png)

The main pipeline is composed of 2 top activities:

- a Lookup activity named "Read Configuration File"
- a ForEach activity named "ForEachYearMonthPair"

> Within the ForEach activity, there is a nested If activity called "If New Or Updated" and if this activity evaluates to `True`, the  "Invoke Ingestion Pipeline" activity is triggered.

Reviewer: ```suggestion
Within the ForEach activity, there is a nested If activity called "If New Or Updated" and if this activity evaluates to `True`, the "Invoke Ingestion Pipeline" activity is triggered.
```

# Nested pipelines

> The following example, exemplifies the utilization of nested pipelines including control activities like ForEach and If.

Reviewer: Perhaps we can start this page with a short description of what the whole pipeline is trying to achieve. Otherwise the reader has to reverse engineer that from the pipeline and description itself and I am struggling to reverse engineer it myself.

Example: `This pipeline is an example that reads the configuration of a lakehouse, then checks which files are newer or updated and then only downloads those files, uploads them to ADLS and eventually into the Lakehouse.` <- Probably incorrect

Reviewer: Thanks, that helps!

![Nested Pipelines](pl_main_pipeline.png)

The main pipeline is composed of 2 top activities:

> - a lookup activity named "Read Configuration File

Reviewer: ```suggestion
- a Lookup activity named "Read Configuration File
```

Author: I've also added the closing quote that was missing.

> - a ForEach activity names "ForEach"

Reviewer: ```suggestion
- a ForEach activity named "ForEach"
```

Author: Typo corrected and I also I've renamed the ForEach activity to "ForEachYearMonthPair" as per your suggestion on the other comment as well.

> Within the ForEach activity, there is a nested If activity called "If New Or Updated" and if this activity evaluates to True, a Invoke Pipeline activity named "Invoke Ingestion Pipeline" is triggered.

Reviewer: ```suggestion
Within the ForEach activity, there is a nested If activity called "If New Or Updated" and if this activity evaluates to True, an Invoke Pipeline activity named "Invoke Ingestion Pipeline" is triggered.
```

Author: Decided to simplify the sentence. Please check if it is OK now!

The Ingestion pipeline is composed also by two activities:

- a Copy activity named "Copy NYCData to ADLS" that copies from the Web to an ADLS Gen2 storage,

> - a Copy activity named "CopyNYCData to Lakehouse" that copies from the ADLS Gen2 storage to the Lakehouse.

Reviewer: ```suggestion
- a Copy activity named "Copy NYCData to Lakehouse" that copies from the ADLS Gen2 storage to the Lakehouse.
```

Reviewer: activity already contains the space

Author: I've renamed this ones based on the feedback from the previous comment. Now they are called "Copy NYCData from Web to ADLS" and "Copy NYCData from ADLS to Lakehouse" and corrected the typo.

![Ingestion Pipeline](pl_ingestion_pipeline.png)

## Testing the pl_main activities

The file that contains the individual activity tests for the pl_main_pipeline is located [here](./fabric/tests/test_pl_main_activity.py) and it contains three individual activity tests:

- test_activity_read_configuration_file, which is responsible to test if the json configuration file is correctly located and defined.

> - test_activity_if, responsible to check id the if condition expression is well formed in order to decide wheather or not new or updated data should be ingested.

Reviewer: ```suggestion
- test_activity_if, responsible to check if the if condition expression is well formed in order to decide whether new or updated data should be ingested.
```

Author: Rewrote it to add more clarity. Also added the new case for the if that evaluates to True and for the if that evaluates to False

- test_activity_invoke_pipeline, responsible to check if the nested pipeline is correctly invoked.

## Testing the pl_ingestion activities

The file that contains the individual activity tests for the pl_ingestion_pipeline is located [here](./fabric/tests/test_pl_ingestion_activity.py) and it contains two individual activity test:

> - test_copy_nyc_data_to_adls2, this test is responsible to test if the source path is correct and id the target path where the data should land in the datalake is well formed and within the correct stablished partition norms.

Reviewer: ```suggestion
- test_copy_nyc_data_to_adls2, this test is responsible to test if the source path is correct and if the target path where the data should land in the datalake is well formed and within the correct stablished partition norms.
```

Author: Rewrote it to add more clarity.

- test_copy_nyc_data_to_lakehouse, similar to the previous one but the previous target is now the source and the current target is the Fabric Lakehouse.

## Testing the pl_main pipeline with should_evaluate_child_pipelines=False

> When testing a pipeline and instantiating the framework, there is a flag called: should_evaluate_child_pipelines. When setting this flag to False, it means that the test is just being done on the main pipeline and the child pipeline will not be loaded for evaluation. Check the code below:

Reviewer 2: ```suggestion
When testing a pipeline and instantiating the framework, there is a flag called: should_evaluate_child_pipelines. When setting this flag to `False`, it means that the test is just being done on the main pipeline and the child pipeline will not be loaded for evaluation. Check the code below:
```

Author: I did some more modifications on this two paragraphs - also added the single quotes on other places where I had True and False referenced. Tks!

```python
def test_framework(request: pytest.FixtureRequest) -> TestFramework:
    return TestFramework(
        framework_type=TestFrameworkType.Fabric,
        root_folder_path = os.path.join(Path(__file__).parent.parent),
        should_evaluate_child_pipelines=False
```

> You can find the entire code  [here](./fabric/tests/test_pl_main_pipeline_child_flag_false.py).

Reviewer 2: ```suggestion
You can find the entire code [here](./fabric/tests/test_pl_main_pipeline_child_flag_false.py).
```

Author: Rewrote to "The full example is available here" - sounds better.

In this specific case you can test for the number of activities within the pipeline and respective types for example. Furthermore, you can do similar tests to the ones were done on the individual activity tests.

For this particular pipeline, we will have two activities to evaluate:

- read_configuration_file_activity = next(activities)
- invoke_ingestion_pipeline_activity =  next(activities)

For the control activities, If and ForEach in this specific pipeline, the framework does the evaluation in the background so they won't be available when running the next(activities) function.

## Testing the pl_main pipeline with should_evaluate_child_pipelines=True

Setting the should_evaluate_child_pipelines flag to True, it means that the test is being done on the main pipeline jointly with the child pipeline. Check the code below:
below:

```python
def test_framework(request: pytest.FixtureRequest) -> TestFramework:
    return TestFramework(
        framework_type=TestFrameworkType.Fabric,
        root_folder_path = os.path.join(Path(__file__).parent.parent),
        should_evaluate_child_pipelines=True
```

> You can find the entire code  [here](./fabric/tests/test_pl_main_pipeline_child_flag_true.py).

Reviewer 2: ```suggestion
You can find the entire code [here](./fabric/tests/test_pl_main_pipeline_child_flag_true.py).
```

Author: Same as previous.
